internal/controller/pipelinerun_controller_test.go

```diff
@@ -14,19 +14,138 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// What to test:
+// - check countRunningPipelineRuns
+// - with a queue of 3 pipelineruns, when one finishes one other starts and one stays pending
+
 package controller
 
 import (
+	"strconv"
+
 	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	tektonv1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	local_tekton "github.com/konflux-ci/mintmaker/internal/pkg/tekton"
+	. "github.com/konflux-ci/mintmaker/pkg/common"
 )
 
 var _ = Describe("PipelineRun Controller", func() {
-	Context("When reconciling a resource", func() {
+	Context("When reconciling pipelineruns", func() {
+
+		originalMaxSimultaneousPipelineRuns := MaxSimultaneousPipelineRuns
+
+		_ = BeforeEach(func() {
+
+			MaxSimultaneousPipelineRuns = 2
+
+			for i := range 3 {
+				pplrName := "pplnr" + strconv.Itoa(i)
+				pipelineRunBuilder := local_tekton.NewPipelineRunBuilder(pplrName, MintMakerNamespaceName)
+				pipelinerun, err := pipelineRunBuilder.Build()
+				Expect(err).NotTo(HaveOccurred())
+				k8sClient.Create(ctx, pipelinerun)
+			}
+			Eventually(listPipelineRuns).WithArguments(MintMakerNamespaceName).Should(HaveLen(3))
+		})
+
+		_ = AfterEach(func() {
+			MaxSimultaneousPipelineRuns = originalMaxSimultaneousPipelineRuns
+
+			// Delete pipelineruns, so they don't leak to other tests
+			pipelineRuns := listPipelineRuns(MintMakerNamespaceName)
+			for _, pipelinerun := range pipelineRuns {
+				k8sClient.Delete(ctx, &pipelinerun)
+			}
+		})
+
+		It("should successfully launch new pipelineruns", func() {
+
+			reconciler := PipelineRunReconciler{
+				Client: k8sClient,
+				Scheme: scheme.Scheme,
+			}
+			reconciler.SetupWithManager(k8sManager)
+			reconciler.Reconcile(ctx, ctrl.Request{
+				NamespacedName: types.NamespacedName{Namespace: MintMakerNamespaceName, Name: "pplnr1"},
+			})
```

Comment:
As mentioned, you should start the controller when the test env is started, instead of calling the reconcile in tests.
You should start the controller similar to this:
https://github.com/konflux-ci/mintmaker/blob/refactoring-pipelinerun/internal/controller/suite_test.go#L101

Reply:
Great catch as well! I was not checking the calls here and was missing more errors.
Thanks for finding this, now everything works great!